_posts/2019-12-18-tfhub-0.7.0/tfhub-0.7.0.Rmd

```diff
@@ -0,0 +1,137 @@
+slug: tfhub
+date: 12-18-2019
+categories:
+  - Keras
+output:
+  distill::distill_article:
+    self_contained: false
+preview: images/tfhub.png
+---
+
+
+```{r setup, include=FALSE}
+knitr::opts_chunk$set(echo = TRUE, eval = FALSE)
+```
+
+We are pleased to announce that the first version of [tfhub](https://github.com/rstudio/tfhub) is now on CRAN. tfhub is an R interface to TensorFlow Hub -  library for the publication, discovery, and consumption of reusable parts of machine learning models. A module is a self-contained piece of a TensorFlow graph, along with its weights and assets, that can be reused across different tasks in a process known as transfer learning. 
+
+The CRAN version of tfhub can be installed with:
+
+```{r}
+install.packages("tfhub")
+```
+
+After installing the R package you need to install the TensorFlow Hub python package. You can do it by running:
+
+```{r}
+tfhub::install_tfhub()
+```
+
+> **Note**: If you are using the most recent CRAN release of reticulate (1.14.0), tfhub will be automatically installed when you `library` it for the first time.
+
+## Simple example
+
+The essencial function of tfhub is `layer_hub` which works just like a [keras](https://github.com/rstudio/keras) but allows you to load a complete pre-trained Deep Learning model.
+
+For example you can:
+
+```{r}
+library(tfhub)
+layer_mobilenet <- layer_hub(
+  handle = "https://tfhub.dev/google/tf2-preview/mobilenet_v2/classification/4"
+)
+```
+
+This will download the MobileNet model pre-trained in the ImageNet dataset. tfhub models are cached locally and don't need to be downloaded the next time you use the same model.
+
+You can now use `layer_mobilenet` as an usual Keras layer. For example you can define a model:
+
+```{r}
+library(keras)
+input <- layer_input(shape = c(224, 224, 3))
+output <- layer_mobilenet(input)
+model <- keras_model(input, output)
+summary(model)
+```
+
+```
+Model: "model"
+____________________________________________________________________
+Layer (type)                  Output Shape               Param #    
+====================================================================
+input_2 (InputLayer)          [(None, 224, 224, 3)]      0          
+____________________________________________________________________
+keras_layer_1 (KerasLayer)    (None, 1001)               3540265    
+====================================================================
+Total params: 3,540,265
+Trainable params: 0
+Non-trainable params: 3,540,265
+____________________________________________________________________
+```
+
+This model can now be used to create predict Imagenet labels for an image. For example, let's see the results for the famous Grace Hopper's photo:
+
+![Grace Hopper](images/grace-hopper.jpg)
+
+```{r}
+img <- image_load("images/grace-hopper.jpg", target_size = c(224,224)) %>% 
+  image_to_array()
+img <- img/255
+dim(img) <- c(1, dim(img))
+pred <- predict(model, img)
+imagenet_decode_predictions(pred[,-1,drop=FALSE])
```

Comment:
Not sure why `imagenet_decode_predictions` returns a list, but think I rather do

```r
imagenet_decode_predictions(pred[,-1,drop=FALSE])[[1]]
```

To avoid printing `[[1]]` in the table.